app/src/org/commcare/tasks/DataPullTask.java

```diff
@@ -102,258 +101,296 @@ public DataPullTask(String username, String password,
         this(username, password, server, context, CommCareApplication._().getDataPullRequester());
     }
 
+    // TODO PLM: once this task is refactored into manageable components, it should use the
+    // ManagedAsyncTask pattern of checking for isCancelled() and aborting at safe places.
     @Override
     protected void onCancelled() {
         super.onCancelled();
-        if (wasKeyLoggedIn) {
-            CommCareApplication._().releaseUserResourcesAndServices();
-        }
+        wipeLoginIfItOccurred();
     }
     private HttpRequestEndpoints requestor;
 
     @Override
     protected ResultAndError<PullTaskResult> doTaskBackground(Void... params) {
-        // Don't try to sync if logging out is occuring
         if (!CommCareSessionService.sessionAliveLock.tryLock()) {
-            // TODO PLM: once this task is refactored into manageable
-            // components, it should use the ManagedAsyncTask pattern of
-            // checking for isCancelled() and aborting at safe places.
-            return new ResultAndError<>(PullTaskResult.UNKNOWN_FAILURE, "Cannot sync while a logout is in process");
+            // Don't try to sync if logging out is occurring
+            return new ResultAndError<>(PullTaskResult.UNKNOWN_FAILURE,
+                    "Cannot sync while a logout is in process");
+        }
+        try {
+            return doTaskBackgroundHelper();
+        } finally {
+            CommCareSessionService.sessionAliveLock.unlock();
+        }
+    }
+
+    private ResultAndError<PullTaskResult> doTaskBackgroundHelper() {
+        publishProgress(PROGRESS_STARTED);
+        recordSyncAttemptTime();
+        Logger.log(AndroidLogger.TYPE_USER, "Starting Sync");
+        determineIfLoginNeeded();
+
+        AndroidTransactionParserFactory factory = getTransactionParserFactory();
+        byte[] wrappedEncryptionKey = getEncryptionKey();
+        if (wrappedEncryptionKey == null) {
+            this.publishProgress(PROGRESS_DONE);
+            return new ResultAndError<>(PullTaskResult.UNKNOWN_FAILURE,
+                    "Unable to get or generate encryption key");
+        } else {
+            factory.initUserParser(wrappedEncryptionKey);
         }
 
-        // Wrap in a 'try' to enable a 'finally' close that releases the
-        // sessionAliveLock.
+        if (isCancelled()) {

```

Comment:
Does anyone know if this check can come earlier? In the old code, this came after the initialization of the transaction parser factory and the encryption key so I left it here, but I'm not sure why that's necessary.